Update how CYLC_DIR is used in Cylc

kinow opened the pull request:
#3133 (not sure if this is enough to close the issue).

This pull request removes some places where `CYLC_DIR` was used. For the test-battery, we used a temporary fix pre-setup.py, where the TRAVIS home is being set to CYLC_DIR.

kinow commented:
Functional tests failing, some are not flaky tests. So it still needs some further work :disappointed_relieved:

kinow commented:
Found a compromise, where CYLC_DIR was not removed, just slightly modified, and its usage when searching for binaries was modified too so that it would look for the right path.

However, now Jinja2 filters are failing...

kinow commented:
Hopefully after #3135 everything should just work :+1:

matthewrmshin commented on lib/cylc/network/client.py:
```diff
@@ -239,11 +241,18 @@ def __new__(cls, suite, owner=None, host=None, port=None, timeout=None):
+                cylc_bin_dir = f"{cylc_bin_dir}/"
+
+            if cmd.startswith(cylc_bin_dir):
+                cmd = cmd.replace(cylc_bin_dir, '')
```
I am probably missing the point of the logic here. Are we just saying that if `cmd` is `/path/to/cylc`, it is replaced with `cylc`?

kinow replied:
That's what I understood too. Only `cylc` commands, do the same as `basename` in Shell I think?

matthewrmshin replied:
`os.path.basename`?

kinow replied:
That one too. I was only a bit scared to change this as I am not familiar with this part of the code yet (i.e. not sure what is using this value, and how to test if I change it).

Using `os.path.basename`, I am not sure if the `cmd` isn't something like `/opt/cylc/bin/cylc run --no-detach --verbose five`... in which case the `.replace` would have to stay, in order to keep the arguments (also not sure if basename would work with arguments).

Maybe we can leave it as-is for now or open another issue to see how to improve or document this.

matthewrmshin replied:
Not possible looking at the code. It really can only be a simple `cylc` command.

matthewrmshin commented on lib/cylc/job_file.py:
```diff
@@ -322,6 +322,6 @@ def _write_script(cls, handle, job_conf):
     def _write_epilogue(handle, job_conf):
         """Write epilogue."""
-        handle.write('\n\n. "${CYLC_DIR}/lib/cylc/job.sh"\ncylc__job__main')
+        handle.write('\n\n. "${CYLC_DIR}/job.sh"\ncylc__job__main')
```
Perhaps we should move the job script to something like `bin/cylc-job.sh` - so it is available under the same location as `bin/cylc`. This should help get rid of `CYLC_DIR` fully.

kinow replied:
Sounds good to me. Is there any risk in doing so though? As this file would be put in the system or user `$PATH` alongside the other scripts.

matthewrmshin replied:
It should be fine. It should not be any riskier than, say, running the `cylc message` command.

hjoliver replied:
👍

matthewrmshin commented on lib/cylc/job.sh:
```diff
@@ -36,26 +36,15 @@ cylc__job__main() {
+        if "${CYLC_DEBUG:-false}"; then
+            . "${JOB_INIT_ENV}"
+        else
+            . "${JOB_INIT_ENV}" 1>'/dev/null' 2>&1
```
I would argue that we should get rid of this logic altogether - now that we have `bash -l` in the job script header.

kinow replied:
Passed over my head. I can dig deeper to see if I understand why `bash -l` would replace it... but I trust another reviewer will have better understanding to comment :grimacing: (and promise to lurk and try to learn a bit more Shell)

matthewrmshin replied:
In the past, we would call `bash` without the `--login` (i.e. `-l`). So users need to set up their environment for the job somehow (so it has access to Cylc, etc). Now that the job script calls `bash` with `--login` (i.e. `-l`), we can safely assume that the correct environment is already available as `--login` will invoke the site/user login profiles.

hjoliver replied:
We still have the option to not use a login shell, however (security people seem not to like jobs using login shells...) ... in which case there might still be a requirement for cylc-specific setup??

matthewrmshin replied:
Like to find out the reason why security people are not keen on login shell. After all, users can put login profiles to the script that get sourced here. If anything, the login shell is a lot more standard and can be more secure, if done right.

kinow replied:
I know that at least in some Node.js and PHP apps, it is common to have keys and tokens as environment variables. Some times under global envvars. I think Django may use that as well?

So maybe some people are worried about those variables being written to log files or to the web interface, leaking the credentials... Just guessing.

matthewrmshin replied:
That would be a bad setup. The user would have to have deliberately written the logic to print out the tokens or have xtrace switched on in the wrong place.

They can do all these stuffs in this script as well.

matthewrmshin replied:
A `printenv` command later can also do it.

kinow replied:
I agree. But if instead they disallowed `bash -l`, I think it would not have happened... which means lower risk, from a security point of view.

My preference is for having separate user accounts, possibly even separated boxes (VM, container, etc), and instead of worrying too much about it, simply follow what is in the user manual. So this could be a possible answer to users concerned about security... though not all would accept I suspect...

matthewrmshin commented:
Assuming that we can resolve the issues with the job scripts.... The only issue left would be the location of site `etc/global.rc` and friends?

kinow commented:
>Assuming that we can resolve the issues with the job scripts.... The only issue left would be the location of site etc/global.rc and friends?

Exactly. I spent some good minutes while working on it trying to come up with a solution that would keep the old behaviour, to be simpler to review and merge.

But alas I see no way to keep using the old approach, and even though I had a look at the documentation looking for a place that could adopt these files... I came with no solutions for this  :disappointed_relieved:

matthewrmshin commented:
The closest thing is to look for the location of the `cylc` command and assuming that it lives under `SOME-PREFIX/bin/cylc`. The logic should then search for `global.rc` from `SOME-PREFIX/etc/cylc/VERSION/global.rc` and `$HOME/.cylc/VERSION/global.rc`.

hjoliver commented:
> The closest thing is to look for the location of the `cylc` command and assuming that it lives under `SOME-PREFIX/bin/cylc`. The logic should then search for `global.rc` from `SOME-PREFIX/etc/cylc/VERSION/global.rc` and `$HOME/.cylc/VERSION/global.rc`.

That doesn't seem unreasonable, but it's not exactly standard is is?  That's why I was wondering (Riot convo) if we should use `/etc` for config files (or and equivalent path under `$HOME/etc` for local user installs....

matthewrmshin commented:
Just thinking... If we are going to deploy using a package manager, and that package managers normally allow you to change the installation *prefix*... Then my comment above is probably just the general case of your suggestion.

I am still not 100% sure how to deal with configuration files of individual versions. I suppose we'll create a local deployment package that:
* Depends on the version of Cylc to deploy.
* Contains the site configuration files.
* Will copy the site configuration files to live under `${PREFIX}/etc`.

kinow commented:
@matthewrmshin 

>The closest thing is to look for the location of the cylc command and assuming that it lives under SOME-PREFIX/bin/cylc. The logic should then search for global.rc from SOME-PREFIX/etc/cylc/VERSION/global.rc and $HOME/.cylc/VERSION/global.rc.

That is the problem that I am having right now. In virtual environments, the `cylc` executable gets put in `venv/bin/cylc`, because it was `scripts` entry point in setuptools.

While the package data in setuptools goes to `venv/lib/python/site-packages/cylc`.

Upgrading Python versions of a virtual environment require - as far as I know - to create a new virtual environment. Which means that you would need to move all your files from the previous location to the new one. I think it would be harder to manage the configuration this way :disappointed: 

Installing with `pip` (or `conda` or `rpm`) put Cylc files in different locations too, and the scripts may not always be in the same directory structure as the rest of the code.

kinow commented:
>I am still not 100% sure how to deal with configuration files of individual versions. I suppose we'll create a local deployment package that

What if we added a configuration to `global.rc`, an array with directories to replace CYLC_DIR?

matthewrmshin commented:
Yes, it will have to be a search path array, but what should be in it?
* /etc/cylc/cylc-flow-VERSION/
* PREFIX/etc/cylc/cylc-flow-VERSION/
* (What else?)
* ~/.cylc/cylc-flow-VERSION/

hjoliver commented:
What about no automatic install of config files?  Just tell users where to put them if they need to override built-in defaults, in some "standard" cylc-version-specific locations under `/etc`, or `<venv>/etc`...?

I think the only config files we provide "out of the box" are the editor syntax files, and we don't currently install those to their editor-specific locations anyway.

hjoliver commented:
Distutils docs suggest using `data_files` for config files.  Relative to installation prefix. https://docs.python.org/3/distutils/setupscript.html#installing-additional-files

On that basis, maybe we should stick with `PREFIX/etc/` ?  And we could put example config files (all items commented out) there.

hjoliver commented:
(And extend search path to handle conda etc. later, when we come to that).

hjoliver commented:
Advantage of `/etc/cylc-flow-VERSION/` (and `$HOME/.cylc-flow-VERSION/`) is you don't have to re-do (or copy) conf files for new installs in new virtual envs.  (I suppose)

kinow commented:
>What about no automatic install of config files? Just tell users where to put them if they need to override built-in defaults, in some "standard" cylc-version-specific locations under /etc, or <venv>/etc...?

That should be OK too I think. Many other tools provide the standard configuration and you are able to tweak it.

>Distutils docs suggest using data_files for config files. Relative to installation prefix. https://docs.python.org/3/distutils/setupscript.html#installing-additional-files
>
>On that basis, maybe we should stick with PREFIX/etc/ ? And we could put example config files (all items commented out) there.

If I understand it correctly, then `PREFIX` be something like `/usr/lib/python3.7/lib/site-packages/cylc-flow/`. In this case, I think system admins would probably prefer to have the configuration file under `/etc/`.

>Advantage of /etc/cylc-flow-VERSION/ (and $HOME/.cylc-flow-VERSION/) is you don't have to re-do (or copy) conf files for new installs in new virtual envs. (I suppose)

Yup. I'm now pending more in favour of `/etc/cylc...`. The `-VERSION` looks a bit different, but I think having `/etc/cylc-8.0a1`, `/etc/cylc-8.0a2`, etc, should be OK? In my `/etc/` right now the only app with two folders like that is gtk, with `/etc/gtk2` and `/etc/gtk3`.

But I think you are thinking ahead when users want to use the global configuration while running multiple versions of Cylc in the same computer. So I don't see another way of having a global configuration somewhere without the version included :confused:

hjoliver commented:
> But I think you are thinking ahead when users want to use the global configuration while running multiple versions of Cylc in the same computer. So I don't see another way of having a global configuration somewhere without the version included

Yes it is quite common to have multiple Cylc versions in use at the same time - we can't expect everyone to upgrade their suites at once, especially long-running suites that keep on cycling even when their owners are away for instance...

hjoliver commented:
It is possible to have config files that are both forward and backward compatible, if the parser just ignores settings that it doesn't understand - but that makes genuine errors hard to detect.

matthewrmshin commented:
It is also worth noting that the system admins who deploy Cylc on site may not be the same people who have full root access.

hjoliver commented:
I'm now running into this problem myself (new desktop box at work, trying to use a proper virtualenv with cylc/cylc-flow for the first time).

kinow commented:
@hjoliver 

>I'm now running into this problem myself (new desktop box at work, trying to use a proper virtualenv with cylc/cylc-flow for the first time).

There is a workaround for development, so that that does not block you (I am using it, otherwise would be pushing more to get it fixed :grimacing: ).

Run `pip install -e .` instead of `pip install .`. The former will install the editable version. Which is done by simply adding a _link_, that says `cylc-flow` has a package `cylc` that is at `/home/user/something/something/cylc-flow/lib`.

This means that `cylc $command` should work, because the `cylc` utility will be found at `/home/user/something/something/cylc-flow/bin/cylc`. So when it tries to find `../lib/cylc`, for example, it works. Simply because the directory structure matches what `CYLC_DIR` was expected to have :+1: 

Just in case that helps you to keep using the virtual environment for now

hjoliver commented:
Huh, I've just discovered, at home on my VirtualBox VM a suite with local background tasks works fine after `pip install .` (not "editable") ... which is surprising .. jobs were failing unable to find `job.sh` on Centos at work.

hjoliver commented:
Correction, it works fine at work too, so long as I do `pip install -e` and not `python setup.py install`, in the venv.

kinow commented:
>Huh, I've just discovered, at home on my VirtualBox VM a suite with local background tasks works fine after pip install . (not "editable") ... which is surprising .. jobs were failing unable to find job.sh on Centos at work.

I had the same thing at home, until I realized in my `$PATH` I had added the `usr/bin/cylc` bash script. It was in `$HOME/bin/cylc`. So I removed it, made sure there was no `cylc` in `pip list | grep cylc`, and installed again.

Then it failed in my virtual environment. So maybe you have another Cylc installation, or that `cylc` bash script?

hjoliver commented:
@kinow - I need to go out for an hour or so at 11:15.  If you have time to rebase this before your lesson, I think I actually have time to work on getting rid of CYLC_DIR entirely, this afternoon, on this branch. (But don't worry if no time).

kinow commented:
Sure, will do. Thanks for reviewing the Jinja2 + namespace PR! :tada:

kinow commented:
Rebased, fixed conflicts (simply love git), and now Travis-CI is building: https://travis-ci.org/kinow/cylc-flow/builds/530033205

Unit tests passed, waiting for functional tests now.

kinow commented:
Had to do one more adjustment for the `cylc/flow/` and `job.sh` work. Running a suite locally with this branch now works. So once again, up to Travis now: https://travis-ci.org/kinow/cylc-flow/builds/530056022

kinow commented:
Hmmm, failing because the `COPYING` file does not exist in `cylc/flow`. This file is not even in `cylc/`, in my virtual environment it is on `cylc_flow_8....a1/COPYING`. Let me think how to fix this one.

hjoliver commented:
@matthewrmshin - I'm part way through completely removing CYLC_DIR on top of this branch.  But that is sufficiently complicated that I think we should merge this PR first, and then I'll do the follow-up.  OK?.

kinow commented:
>But that is sufficiently complicated

Definitely agree on this part! :)

matthewrmshin commented:
@kinow The long output test in the test battery uses the `COPYING` file as a convenient source of a stable and *big enough* file that would cause the output pipe buffer problem in a subprocess.

I guess `COPYING` is no available in the installed environment? We can do either:
* Use any file of similar size for the test instead.
* Install the `COPYING` file somewhere?

@hjoliver I am happy to merge this branch first - as soon as the above test can be made to pass.

kinow commented:
Alright, I will have to ask for help with the functional test. I swear I tried :disappointed_relieved: 

So the current `COPYING` file is not included. Instead I thought about a simple solution of copyng `COPYING` to `tests/events/47-long-output/COPYING`.

The line that needs changing has:

```
succeeded handler = cat "${CYLC_DIR}/COPYING" "${CYLC_DIR}/COPYING" "${CYLC_DIR}/COPYING" >&2 && echo
```

So I tried replacing `${CYLC_DIR}` by `TEST_NAME_BASE`, `${TEST_SOURCE_DIR}/${TEST_NAME_BASE}`, even tried using Jinja2 (i.e. adding the `#!Jinja2`) and passing the variables via `--set=BLABLA=BLABLA`, but it still did not find the files :disappointed: 

Tried looking at other tests, and also searched the CUG, but had no luck. Then I noticed that I had spent about half the time put into `CYLC_DIR` change just trying to figure out how to fix this test and decided to ask for a little help :slightly_smiling_face: Please.

hjoliver commented:
@kinow @matthewrmshin - I'll fix that test on my follow-up branch, since I'm dealing with "package data" in a big way ... now know how to install files like that and how to access them.

hjoliver commented:
(I may be being overly optimistic if the comments above are anything to go by ... but we'll see!)

hjoliver commented:
(I think accepting a PR with a failing test is acceptable in this case 😬 )

hjoliver commented:
(or I could have a crack at it on this branch... taking a look now).

kinow commented:
Found a way to make that test pass locally @hjoliver . Here's what I did in my working copy (did not have to have a `suite.rc`, but half my tests failed twice because I forgot to update the other suite in the test :angry:):

```diff
diff --git a/tests/events/47-long-output.t b/tests/events/47-long-output.t
index 13b6335f6..b5ee5a160 100755
--- a/tests/events/47-long-output.t
+++ b/tests/events/47-long-output.t
@@ -29,17 +29,9 @@ process pool timeout = PT10S" ""
 
 # Long STDOUT output
 
-init_suite "${TEST_NAME_BASE}" <<'__SUITERC__'
-[cylc]
-[scheduling]
-    [[dependencies]]
-        graph = t1
-[runtime]
-    [[t1]]
-        script = true
-        [[[events]]]
-            succeeded handler = cat "${CYLC_DIR}/COPYING" "${CYLC_DIR}/COPYING" "${CYLC_DIR}/COPYING" && echo
-__SUITERC__
+install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
+
+cp "${TEST_SOURCE_DIR}/${TEST_NAME_BASE}/COPYING" "${SUITE_RUN_DIR}"
 
 run_ok "${TEST_NAME_BASE}-validate" cylc validate "${SUITE_NAME}"
 
@@ -62,17 +54,9 @@ purge_suite "${SUITE_NAME}"
 
 # REPEAT: Long STDERR output
 
-init_suite "${TEST_NAME_BASE}" <<'__SUITERC__'
-[cylc]
-[scheduling]
-    [[dependencies]]
-        graph = t1
-[runtime]
-    [[t1]]
-        script = true
-        [[[events]]]
-            succeeded handler = cat "${CYLC_DIR}/COPYING" "${CYLC_DIR}/COPYING" "${CYLC_DIR}/COPYING" >&2 && echo
-__SUITERC__
+install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
+
+cp "${TEST_SOURCE_DIR}/${TEST_NAME_BASE}/COPYING" "${SUITE_RUN_DIR}"
 
 run_ok "${TEST_NAME_BASE}-validate" cylc validate "${SUITE_NAME}"
 
diff --git a/tests/events/47-long-output/COPYING b/tests/events/47-long-output/COPYING
new file mode 100644
index 000000000..94a9ed024
--- /dev/null
+++ b/tests/events/47-long-output/COPYING
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
+                       Version 3, 29 June 2007
+
+ Copyright (C) 2007 Free Software Foundation, Inc. <http://fsf.org/>
+ Everyone is permitted to copy and distribute verbatim copies
+ of this license document, but changing it is not allowed.
+
+                            Preamble
+
+  The GNU General Public License is a free, copyleft license for
+software and other kinds of works.
+
+  The licenses for most software and other practical works are designed
+to take away your freedom to share and change the works.  By contrast,
+the GNU General Public License is intended to guarantee your freedom to
+share and change all versions of a program--to make sure it remains free
+software for all its users.  We, the Free Software Foundation, use the
+GNU General Public License for most of our software; it applies also to
+any other work released this way by its authors.  You can apply it to
+your programs, too.
+
+  When we speak of free software, we are referring to freedom, not
+price.  Our General Public Licenses are designed to make sure that you
+have the freedom to distribute copies of free software (and charge for
+them if you wish), that you receive source code or can get it if you
+want it, that you can change the software or use pieces of it in new
+free programs, and that you know you can do these things.
+
+  To protect your rights, we need to prevent others from denying you
+these rights or asking you to surrender the rights.  Therefore, you have
+certain responsibilities if you distribute copies of the software, or if
+you modify it: responsibilities to respect the freedom of others.
+
+  For example, if you distribute copies of such a program, whether
+gratis or for a fee, you must pass on to the recipients the same
+freedoms that you received.  You must make sure that they, too, receive
+or can get the source code.  And you must show them these terms so they
+know their rights.
+
+  Developers that use the GNU GPL protect your rights with two steps:
+(1) assert copyright on the software, and (2) offer you this License
+giving you legal permission to copy, distribute and/or modify it.
+
+  For the developers' and authors' protection, the GPL clearly explains
+that there is no warranty for this free software.  For both users' and
+authors' sake, the GPL requires that modified versions be marked as
+changed, so that their problems will not be attributed erroneously to
+authors of previous versions.
+
+  Some devices are designed to deny users access to install or run
+modified versions of the software inside them, although the manufacturer
+can do so.  This is fundamentally incompatible with the aim of
+protecting users' freedom to change the software.  The systematic
+pattern of such abuse occurs in the area of products for individuals to
+use, which is precisely where it is most unacceptable.  Therefore, we
+have designed this version of the GPL to prohibit the practice for those
+products.  If such problems arise substantially in other domains, we
+stand ready to extend this provision to those domains in future versions
+of the GPL, as needed to protect the freedom of users.
+
+  Finally, every program is threatened constantly by software patents.
+States should not allow patents to restrict development and use of
+software on general-purpose computers, but in those that do, we wish to
+avoid the special danger that patents applied to a free program could
+make it effectively proprietary.  To prevent this, the GPL assures that
+patents cannot be used to render the program non-free.
+
+  The precise terms and conditions for copying, distribution and
+modification follow.
+
+                       TERMS AND CONDITIONS
+
+  0. Definitions.
+
+  "This License" refers to version 3 of the GNU General Public License.
+
+  "Copyright" also means copyright-like laws that apply to other kinds of
+works, such as semiconductor masks.
+
+  "The Program" refers to any copyrightable work licensed under this
+License.  Each licensee is addressed as "you".  "Licensees" and
+"recipients" may be individuals or organizations.
+
+  To "modify" a work means to copy from or adapt all or part of the work
+in a fashion requiring copyright permission, other than the making of an
+exact copy.  The resulting work is called a "modified version" of the
+earlier work or a work "based on" the earlier work.
+
+  A "covered work" means either the unmodified Program or a work based
+on the Program.
+
+  To "propagate" a work means to do anything with it that, without
+permission, would make you directly or secondarily liable for
+infringement under applicable copyright law, except executing it on a
+computer or modifying a private copy.  Propagation includes copying,
+distribution (with or without modification), making available to the
+public, and in some countries other activities as well.
+
+  To "convey" a work means any kind of propagation that enables other
+parties to make or receive copies.  Mere interaction with a user through
+a computer network, with no transfer of a copy, is not conveying.
+
+  An interactive user interface displays "Appropriate Legal Notices"
+to the extent that it includes a convenient and prominently visible
+feature that (1) displays an appropriate copyright notice, and (2)
+tells the user that there is no warranty for the work (except to the
+extent that warranties are provided), that licensees may convey the
+work under this License, and how to view a copy of this License.  If
+the interface presents a list of user commands or options, such as a
+menu, a prominent item in the list meets this criterion.
+
+  1. Source Code.
+
+  The "source code" for a work means the preferred form of the work
+for making modifications to it.  "Object code" means any non-source
+form of a work.
+
+  A "Standard Interface" means an interface that either is an official
+standard defined by a recognized standards body, or, in the case of
+interfaces specified for a particular programming language, one that
+is widely used among developers working in that language.
+
+  The "System Libraries" of an executable work include anything, other
+than the work as a whole, that (a) is included in the normal form of
+packaging a Major Component, but which is not part of that Major
+Component, and (b) serves only to enable use of the work with that
+Major Component, or to implement a Standard Interface for which an
+implementation is available to the public in source code form.  A
+"Major Component", in this context, means a major essential component
+(kernel, window system, and so on) of the specific operating system
+(if any) on which the executable work runs, or a compiler used to
+produce the work, or an object code interpreter used to run it.
+
+  The "Corresponding Source" for a work in object code form means all
+the source code needed to generate, install, and (for an executable
+work) run the object code and to modify the work, including scripts to
+control those activities.  However, it does not include the work's
+System Libraries, or general-purpose tools or generally available free
+programs which are used unmodified in performing those activities but
+which are not part of the work.  For example, Corresponding Source
+includes interface definition files associated with source files for
+the work, and the source code for shared libraries and dynamically
+linked subprograms that the work is specifically designed to require,
+such as by intimate data communication or control flow between those
+subprograms and other parts of the work.
+
+  The Corresponding Source need not include anything that users
+can regenerate automatically from other parts of the Corresponding
+Source.
+
+  The Corresponding Source for a work in source code form is that
+same work.
+
+  2. Basic Permissions.
+
+  All rights granted under this License are granted for the term of
+copyright on the Program, and are irrevocable provided the stated
+conditions are met.  This License explicitly affirms your unlimited
+permission to run the unmodified Program.  The output from running a
+covered work is covered by this License only if the output, given its
+content, constitutes a covered work.  This License acknowledges your
+rights of fair use or other equivalent, as provided by copyright law.
+
+  You may make, run and propagate covered works that you do not
+convey, without conditions so long as your license otherwise remains
+in force.  You may convey covered works to others for the sole purpose
+of having them make modifications exclusively for you, or provide you
+with facilities for running those works, provided that you comply with
+the terms of this License in conveying all material for which you do
+not control copyright.  Those thus making or running the covered works
+for you must do so exclusively on your behalf, under your direction
+and control, on terms that prohibit them from making any copies of
+your copyrighted material outside their relationship with you.
+
+  Conveying under any other circumstances is permitted solely under
+the conditions stated below.  Sublicensing is not allowed; section 10
+makes it unnecessary.
+
+  3. Protecting Users' Legal Rights From Anti-Circumvention Law.
+
+  No covered work shall be deemed part of an effective technological
+measure under any applicable law fulfilling obligations under article
+11 of the WIPO copyright treaty adopted on 20 December 1996, or
+similar laws prohibiting or restricting circumvention of such
+measures.
+
+  When you convey a covered work, you waive any legal power to forbid
+circumvention of technological measures to the extent such circumvention
+is effected by exercising rights under this License with respect to
+the covered work, and you disclaim any intention to limit operation or
+modification of the work as a means of enforcing, against the work's
+users, your or third parties' legal rights to forbid circumvention of
+technological measures.
+
+  4. Conveying Verbatim Copies.
+
+  You may convey verbatim copies of the Program's source code as you
+receive it, in any medium, provided that you conspicuously and
+appropriately publish on each copy an appropriate copyright notice;
+keep intact all notices stating that this License and any
+non-permissive terms added in accord with section 7 apply to the code;
+keep intact all notices of the absence of any warranty; and give all
+recipients a copy of this License along with the Program.
+
+  You may charge any price or no price for each copy that you convey,
+and you may offer support or warranty protection for a fee.
+
+  5. Conveying Modified Source Versions.
+
+  You may convey a work based on the Program, or the modifications to
+produce it from the Program, in the form of source code under the
+terms of section 4, provided that you also meet all of these conditions:
+
+    a) The work must carry prominent notices stating that you modified
+    it, and giving a relevant date.
+
+    b) The work must carry prominent notices stating that it is
+    released under this License and any conditions added under section
+    7.  This requirement modifies the requirement in section 4 to
+    "keep intact all notices".
+
+    c) You must license the entire work, as a whole, under this
+    License to anyone who comes into possession of a copy.  This
+    License will therefore apply, along with any applicable section 7
+    additional terms, to the whole of the work, and all its parts,
+    regardless of how they are packaged.  This License gives no
+    permission to license the work in any other way, but it does not
+    invalidate such permission if you have separately received it.
+
+    d) If the work has interactive user interfaces, each must display
+    Appropriate Legal Notices; however, if the Program has interactive
+    interfaces that do not display Appropriate Legal Notices, your
+    work need not make them do so.
+
+  A compilation of a covered work with other separate and independent
+works, which are not by their nature extensions of the covered work,
+and which are not combined with it such as to form a larger program,
+in or on a volume of a storage or distribution medium, is called an
+"aggregate" if the compilation and its resulting copyright are not
+used to limit the access or legal rights of the compilation's users
+beyond what the individual works permit.  Inclusion of a covered work
+in an aggregate does not cause this License to apply to the other
+parts of the aggregate.
+
+  6. Conveying Non-Source Forms.
+
+  You may convey a covered work in object code form under the terms
+of sections 4 and 5, provided that you also convey the
+machine-readable Corresponding Source under the terms of this License,
+in one of these ways:
+
+    a) Convey the object code in, or embodied in, a physical product
+    (including a physical distribution medium), accompanied by the
+    Corresponding Source fixed on a durable physical medium
+    customarily used for software interchange.
+
+    b) Convey the object code in, or embodied in, a physical product
+    (including a physical distribution medium), accompanied by a
+    written offer, valid for at least three years and valid for as
+    long as you offer spare parts or customer support for that product
+    model, to give anyone who possesses the object code either (1) a
+    copy of the Corresponding Source for all the software in the
+    product that is covered by this License, on a durable physical
+    medium customarily used for software interchange, for a price no
+    more than your reasonable cost of physically performing this
+    conveying of source, or (2) access to copy the
+    Corresponding Source from a network server at no charge.
+
+    c) Convey individual copies of the object code with a copy of the
+    written offer to provide the Corresponding Source.  This
+    alternative is allowed only occasionally and noncommercially, and
+    only if you received the object code with such an offer, in accord
+    with subsection 6b.
+
+    d) Convey the object code by offering access from a designated
+    place (gratis or for a charge), and offer equivalent access to the
+    Corresponding Source in the same way through the same place at no
+    further charge.  You need not require recipients to copy the
+    Corresponding Source along with the object code.  If the place to
+    copy the object code is a network server, the Corresponding Source
+    may be on a different server (operated by you or a third party)
+    that supports equivalent copying facilities, provided you maintain
+    clear directions next to the object code saying where to find the
+    Corresponding Source.  Regardless of what server hosts the
+    Corresponding Source, you remain obligated to ensure that it is
+    available for as long as needed to satisfy these requirements.
+
+    e) Convey the object code using peer-to-peer transmission, provided
+    you inform other peers where the object code and Corresponding
+    Source of the work are being offered to the general public at no
+    charge under subsection 6d.
+
+  A separable portion of the object code, whose source code is excluded
+from the Corresponding Source as a System Library, need not be
+included in conveying the object code work.
+
+  A "User Product" is either (1) a "consumer product", which means any
+tangible personal property which is normally used for personal, family,
+or household purposes, or (2) anything designed or sold for incorporation
+into a dwelling.  In determining whether a product is a consumer product,
+doubtful cases shall be resolved in favor of coverage.  For a particular
+product received by a particular user, "normally used" refers to a
+typical or common use of that class of product, regardless of the status
+of the particular user or of the way in which the particular user
+actually uses, or expects or is expected to use, the product.  A product
+is a consumer product regardless of whether the product has substantial
+commercial, industrial or non-consumer uses, unless such uses represent
+the only significant mode of use of the product.
+
+  "Installation Information" for a User Product means any methods,
+procedures, authorization keys, or other information required to install
+and execute modified versions of a covered work in that User Product from
+a modified version of its Corresponding Source.  The information must
+suffice to ensure that the continued functioning of the modified object
+code is in no case prevented or interfered with solely because
+modification has been made.
+
+  If you convey an object code work under this section in, or with, or
+specifically for use in, a User Product, and the conveying occurs as
+part of a transaction in which the right of possession and use of the
+User Product is transferred to the recipient in perpetuity or for a
+fixed term (regardless of how the transaction is characterized), the
+Corresponding Source conveyed under this section must be accompanied
+by the Installation Information.  But this requirement does not apply
+if neither you nor any third party retains the ability to install
+modified object code on the User Product (for example, the work has
+been installed in ROM).
+
+  The requirement to provide Installation Information does not include a
+requirement to continue to provide support service, warranty, or updates
+for a work that has been modified or installed by the recipient, or for
+the User Product in which it has been modified or installed.  Access to a
+network may be denied when the modification itself materially and
+adversely affects the operation of the network or violates the rules and
+protocols for communication across the network.
+
+  Corresponding Source conveyed, and Installation Information provided,
+in accord with this section must be in a format that is publicly
+documented (and with an implementation available to the public in
+source code form), and must require no special password or key for
+unpacking, reading or copying.
+
+  7. Additional Terms.
+
+  "Additional permissions" are terms that supplement the terms of this
+License by making exceptions from one or more of its conditions.
+Additional permissions that are applicable to the entire Program shall
+be treated as though they were included in this License, to the extent
+that they are valid under applicable law.  If additional permissions
+apply only to part of the Program, that part may be used separately
+under those permissions, but the entire Program remains governed by
+this License without regard to the additional permissions.
+
+  When you convey a copy of a covered work, you may at your option
+remove any additional permissions from that copy, or from any part of
+it.  (Additional permissions may be written to require their own
+removal in certain cases when you modify the work.)  You may place
+additional permissions on material, added by you to a covered work,
+for which you have or can give appropriate copyright permission.
+
+  Notwithstanding any other provision of this License, for material you
+add to a covered work, you may (if authorized by the copyright holders of
+that material) supplement the terms of this License with terms:
+
+    a) Disclaiming warranty or limiting liability differently from the
+    terms of sections 15 and 16 of this License; or
+
+    b) Requiring preservation of specified reasonable legal notices or
+    author attributions in that material or in the Appropriate Legal
+    Notices displayed by works containing it; or
+
+    c) Prohibiting misrepresentation of the origin of that material, or
+    requiring that modified versions of such material be marked in
+    reasonable ways as different from the original version; or
+
+    d) Limiting the use for publicity purposes of names of licensors or
+    authors of the material; or
+
+    e) Declining to grant rights under trademark law for use of some
+    trade names, trademarks, or service marks; or
+
+    f) Requiring indemnification of licensors and authors of that
+    material by anyone who conveys the material (or modified versions of
+    it) with contractual assumptions of liability to the recipient, for
+    any liability that these contractual assumptions directly impose on
+    those licensors and authors.
+
+  All other non-permissive additional terms are considered "further
+restrictions" within the meaning of section 10.  If the Program as you
+received it, or any part of it, contains a notice stating that it is
+governed by this License along with a term that is a further
+restriction, you may remove that term.  If a license document contains
+a further restriction but permits relicensing or conveying under this
+License, you may add to a covered work material governed by the terms
+of that license document, provided that the further restriction does
+not survive such relicensing or conveying.
+
+  If you add terms to a covered work in accord with this section, you
+must place, in the relevant source files, a statement of the
+additional terms that apply to those files, or a notice indicating
+where to find the applicable terms.
+
+  Additional terms, permissive or non-permissive, may be stated in the
+form of a separately written license, or stated as exceptions;
+the above requirements apply either way.
+
+  8. Termination.
+
+  You may not propagate or modify a covered work except as expressly
+provided under this License.  Any attempt otherwise to propagate or
+modify it is void, and will automatically terminate your rights under
+this License (including any patent licenses granted under the third
+paragraph of section 11).
+
+  However, if you cease all violation of this License, then your
+license from a particular copyright holder is reinstated (a)
+provisionally, unless and until the copyright holder explicitly and
+finally terminates your license, and (b) permanently, if the copyright
+holder fails to notify you of the violation by some reasonable means
+prior to 60 days after the cessation.
+
+  Moreover, your license from a particular copyright holder is
+reinstated permanently if the copyright holder notifies you of the
+violation by some reasonable means, this is the first time you have
+received notice of violation of this License (for any work) from that
+copyright holder, and you cure the violation prior to 30 days after
+your receipt of the notice.
+
+  Termination of your rights under this section does not terminate the
+licenses of parties who have received copies or rights from you under
+this License.  If your rights have been terminated and not permanently
+reinstated, you do not qualify to receive new licenses for the same
+material under section 10.
+
+  9. Acceptance Not Required for Having Copies.
+
+  You are not required to accept this License in order to receive or
+run a copy of the Program.  Ancillary propagation of a covered work
+occurring solely as a consequence of using peer-to-peer transmission
+to receive a copy likewise does not require acceptance.  However,
+nothing other than this License grants you permission to propagate or
+modify any covered work.  These actions infringe copyright if you do
+not accept this License.  Therefore, by modifying or propagating a
+covered work, you indicate your acceptance of this License to do so.
+
+  10. Automatic Licensing of Downstream Recipients.
+
+  Each time you convey a covered work, the recipient automatically
+receives a license from the original licensors, to run, modify and
+propagate that work, subject to this License.  You are not responsible
+for enforcing compliance by third parties with this License.
+
+  An "entity transaction" is a transaction transferring control of an
+organization, or substantially all assets of one, or subdividing an
+organization, or merging organizations.  If propagation of a covered
+work results from an entity transaction, each party to that
+transaction who receives a copy of the work also receives whatever
+licenses to the work the party's predecessor in interest had or could
+give under the previous paragraph, plus a right to possession of the
+Corresponding Source of the work from the predecessor in interest, if
+the predecessor has it or can get it with reasonable efforts.
+
+  You may not impose any further restrictions on the exercise of the
+rights granted or affirmed under this License.  For example, you may
+not impose a license fee, royalty, or other charge for exercise of
+rights granted under this License, and you may not initiate litigation
+(including a cross-claim or counterclaim in a lawsuit) alleging that
+any patent claim is infringed by making, using, selling, offering for
+sale, or importing the Program or any portion of it.
+
+  11. Patents.
+
+  A "contributor" is a copyright holder who authorizes use under this
+License of the Program or a work on which the Program is based.  The
+work thus licensed is called the contributor's "contributor version".
+
+  A contributor's "essential patent claims" are all patent claims
+owned or controlled by the contributor, whether already acquired or
+hereafter acquired, that would be infringed by some manner, permitted
+by this License, of making, using, or selling its contributor version,
+but do not include claims that would be infringed only as a
+consequence of further modification of the contributor version.  For
+purposes of this definition, "control" includes the right to grant
+patent sublicenses in a manner consistent with the requirements of
+this License.
+
+  Each contributor grants you a non-exclusive, worldwide, royalty-free
+patent license under the contributor's essential patent claims, to
+make, use, sell, offer for sale, import and otherwise run, modify and
+propagate the contents of its contributor version.
+
+  In the following three paragraphs, a "patent license" is any express
+agreement or commitment, however denominated, not to enforce a patent
+(such as an express permission to practice a patent or covenant not to
+sue for patent infringement).  To "grant" such a patent license to a
+party means to make such an agreement or commitment not to enforce a
+patent against the party.
+
+  If you convey a covered work, knowingly relying on a patent license,
+and the Corresponding Source of the work is not available for anyone
+to copy, free of charge and under the terms of this License, through a
+publicly available network server or other readily accessible means,
+then you must either (1) cause the Corresponding Source to be so
+available, or (2) arrange to deprive yourself of the benefit of the
+patent license for this particular work, or (3) arrange, in a manner
+consistent with the requirements of this License, to extend the patent
+license to downstream recipients.  "Knowingly relying" means you have
+actual knowledge that, but for the patent license, your conveying the
+covered work in a country, or your recipient's use of the covered work
+in a country, would infringe one or more identifiable patents in that
+country that you have reason to believe are valid.
+
+  If, pursuant to or in connection with a single transaction or
+arrangement, you convey, or propagate by procuring conveyance of, a
+covered work, and grant a patent license to some of the parties
+receiving the covered work authorizing them to use, propagate, modify
+or convey a specific copy of the covered work, then the patent license
+you grant is automatically extended to all recipients of the covered
+work and works based on it.
+
+  A patent license is "discriminatory" if it does not include within
+the scope of its coverage, prohibits the exercise of, or is
+conditioned on the non-exercise of one or more of the rights that are
+specifically granted under this License.  You may not convey a covered
+work if you are a party to an arrangement with a third party that is
+in the business of distributing software, under which you make payment
+to the third party based on the extent of your activity of conveying
+the work, and under which the third party grants, to any of the
+parties who would receive the covered work from you, a discriminatory
+patent license (a) in connection with copies of the covered work
+conveyed by you (or copies made from those copies), or (b) primarily
+for and in connection with specific products or compilations that
+contain the covered work, unless you entered into that arrangement,
+or that patent license was granted, prior to 28 March 2007.
+
+  Nothing in this License shall be construed as excluding or limiting
+any implied license or other defenses to infringement that may
+otherwise be available to you under applicable patent law.
+
+  12. No Surrender of Others' Freedom.
+
+  If conditions are imposed on you (whether by court order, agreement or
+otherwise) that contradict the conditions of this License, they do not
+excuse you from the conditions of this License.  If you cannot convey a
+covered work so as to satisfy simultaneously your obligations under this
+License and any other pertinent obligations, then as a consequence you may
+not convey it at all.  For example, if you agree to terms that obligate you
+to collect a royalty for further conveying from those to whom you convey
+the Program, the only way you could satisfy both those terms and this
+License would be to refrain entirely from conveying the Program.
+
+  13. Use with the GNU Affero General Public License.
+
+  Notwithstanding any other provision of this License, you have
+permission to link or combine any covered work with a work licensed
+under version 3 of the GNU Affero General Public License into a single
+combined work, and to convey the resulting work.  The terms of this
+License will continue to apply to the part which is the covered work,
+but the special requirements of the GNU Affero General Public License,
+section 13, concerning interaction through a network will apply to the
+combination as such.
+
+  14. Revised Versions of this License.
+
+  The Free Software Foundation may publish revised and/or new versions of
+the GNU General Public License from time to time.  Such new versions will
+be similar in spirit to the present version, but may differ in detail to
+address new problems or concerns.
+
+  Each version is given a distinguishing version number.  If the
+Program specifies that a certain numbered version of the GNU General
+Public License "or any later version" applies to it, you have the
+option of following the terms and conditions either of that numbered
+version or of any later version published by the Free Software
+Foundation.  If the Program does not specify a version number of the
+GNU General Public License, you may choose any version ever published
+by the Free Software Foundation.
+
+  If the Program specifies that a proxy can decide which future
+versions of the GNU General Public License can be used, that proxy's
+public statement of acceptance of a version permanently authorizes you
+to choose that version for the Program.
+
+  Later license versions may give you additional or different
+permissions.  However, no additional obligations are imposed on any
+author or copyright holder as a result of your choosing to follow a
+later version.
+
+  15. Disclaimer of Warranty.
+
+  THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY
+APPLICABLE LAW.  EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT
+HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY
+OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO,
+THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM
+IS WITH YOU.  SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF
+ALL NECESSARY SERVICING, REPAIR OR CORRECTION.
+
+  16. Limitation of Liability.
+
+  IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
+WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS
+THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY
+GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE
+USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF
+DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD
+PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS),
+EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
+SUCH DAMAGES.
+
+  17. Interpretation of Sections 15 and 16.
+
+  If the disclaimer of warranty and limitation of liability provided
+above cannot be given local legal effect according to their terms,
+reviewing courts shall apply local law that most closely approximates
+an absolute waiver of all civil liability in connection with the
+Program, unless a warranty or assumption of liability accompanies a
+copy of the Program in return for a fee.
+
+                     END OF TERMS AND CONDITIONS
+
+            How to Apply These Terms to Your New Programs
+
+  If you develop a new program, and you want it to be of the greatest
+possible use to the public, the best way to achieve this is to make it
+free software which everyone can redistribute and change under these terms.
+
+  To do so, attach the following notices to the program.  It is safest
+to attach them to the start of each source file to most effectively
+state the exclusion of warranty; and each file should have at least
+the "copyright" line and a pointer to where the full notice is found.
+
+    <one line to give the program's name and a brief idea of what it does.>
+    Copyright (C) <year>  <name of author>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Also add information on how to contact you by electronic and paper mail.
+
+  If the program does terminal interaction, make it output a short
+notice like this when it starts in an interactive mode:
+
+    <program>  Copyright (C) <year>  <name of author>
+    This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
+    This is free software, and you are welcome to redistribute it
+    under certain conditions; type `show c' for details.
+
+The hypothetical commands `show w' and `show c' should show the appropriate
+parts of the General Public License.  Of course, your program's commands
+might be different; for a GUI interface, you would use an "about box".
+
+  You should also get your employer (if you work as a programmer) or school,
+if any, to sign a "copyright disclaimer" for the program, if necessary.
+For more information on this, and how to apply and follow the GNU GPL, see
+<http://www.gnu.org/licenses/>.
+
+  The GNU General Public License does not permit incorporating your program
+into proprietary programs.  If your program is a subroutine library, you
+may consider it more useful to permit linking proprietary applications with
+the library.  If this is what you want to do, use the GNU Lesser General
+Public License instead of this License.  But first, please read
+<http://www.gnu.org/philosophy/why-not-lgpl.html>.
diff --git a/tests/events/47-long-output/suite.rc b/tests/events/47-long-output/suite.rc
new file mode 100644
index 000000000..daf00c12a
--- /dev/null
+++ b/tests/events/47-long-output/suite.rc
@@ -0,0 +1,9 @@
+[cylc]
+[scheduling]
+    [[dependencies]]
+        graph = t1
+[runtime]
+    [[t1]]
+        script = true
+        [[[events]]]
+            succeeded handler = cat "${CYLC_SUITE_RUN_DIR}/COPYING" "${CYLC_SUITE_RUN_DIR}/COPYING" "${CYLC_SUITE_RUN_DIR}/COPYING" >&2 && echo
```

kinow commented:
TL;DR: copied `COPYING` to the test directory. Then I install the suite, and in sequence copy the `COPYING` to the cylc-run directory of that suite. Then the suite simply `cat`s `${CYLC_SUITE_RUN_DIR}/COPYING`, instead of using `CYLC_DIR`. Not sure if too hacky, or if that's how we are supposed to access test resource files.

hjoliver commented:
Yes I just pushed a simpler fix - two line change in the test script!  (works locally at least; I think it will work on Travis.

kinow commented:
Ha! You see, I knew it! :clap:  Thanks!!!

matthewrmshin commented:
@kinow do you want to press the *Ready for review* button?

kinow commented:
Done @matthewrmshin . Close/open to trigger Travis.

hjoliver commented on tests/events/47-long-output.t:
```diff
@@ -62,7 +62,7 @@ purge_suite "${SUITE_NAME}"
 # REPEAT: Long STDERR output
 
-init_suite "${TEST_NAME_BASE}" <<'__SUITERC__'
+init_suite "${TEST_NAME_BASE}" <<__SUITERC__
```
Unquoted heredoc delimiter lets `$CYLC_DIR` get interpolated by the shell immediately, so that actual directory path to the top level of the repository (not the installed library) gets written to the suite definition, rather than the variable.

This fixes the problem because `$CYLC_DIR` is defined wrongly (for this purpose) in `cylc/flow/__init__.py`... 

`$CYLC_DIR` was always the top level of the repository, not of the Python library...

kinow replied:
Thanks for the explanation!!! Will try to remember that next time I have a similar problem with another test :+1:

hjoliver replied:
(And in the Travis build, CYLC_DIR gets set to TRAVIS_BUILD_DIR (i.e. repository top level, where COPYING is) by the test battery command).

hjoliver commented:
The test battery is not part of the installed package, and our Travis build runs out of the repository, so it has access to `COPYING` at the top level of the repository ... so that file doesn't actually need to be installed with the package.

hjoliver commented:
`tests/events/47-long-output.t` just passed in chunk 2  .... phew 😅

kinow commented:
`00-background.t` failed there, in `2/4`. But it's in the list of flaky tests. So kicking Travis.

hjoliver commented:
Thanks, pretty sure it's a spurious failure, passes locally.

kinow commented:
It would take me a long time to be able to write those two lines :) thanks a lot!

hjoliver commented:
I'll take the blame, for the horror that `$CYLC_DIR` had become!
